File: lab_02/winograd.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define A_ROWS 4
#define A_COLS 4

#define B_COLS 2


void winograd(  int** a, int r_a, int c_a , 
                int** b, int r_b, int c_b, 
                int** c, int r_c, int c_c);

int** allocateM(size_t n, size_t m);
int randM(int** a, size_t r, size_t c);
void printM( int** a, size_t len1, size_t len2 );
void freeM(int** a, size_t r);


int main()
{
    srand(time(NULL));
    int** a = allocateM(A_ROWS, A_COLS);
    randM(a, A_ROWS, A_COLS);

    int** b = allocateM(A_ROWS, B_COLS);
    randM(b, A_ROWS, B_COLS);

    int** c = allocateM(A_ROWS, B_COLS);

    if(a && b && c)
        winograd(   a, A_ROWS, A_COLS,
                    b, A_ROWS, B_COLS,
                    c, A_ROWS, B_COLS);

    printM(a, A_ROWS, A_COLS);
    printf("\n");
    printM(b, A_ROWS, B_COLS);
    printf("\n");
    printM(c, A_ROWS, B_COLS);

    freeM(a, A_ROWS);
    freeM(b, A_ROWS);
    freeM(c, A_ROWS);

    return 1;
}

void winograd(  int** a, int r_a, int c_a , 
                int** b, int r_b, int c_b, 
                int** c, int r_c, int c_c)
{
    int d = c_a / 2;

    int* rows = (int*)malloc(sizeof(int) * r_a);
    for(int i = 0; i < r_a; i++) {
        rows[i] = 0;
        rows[i] += a[i][0] * a[i][1];
        for(int j = 1; j < d; j++)
            rows[i] += a[i][2*j] * a[i][2*j+1];
    }

    int* columns = (int*)malloc(sizeof(int) * c_b);
    for(int i = 0; i < c_b; i++) {
        columns[i] = 0;
        columns[i] += b[0][i]*b[1][i];
        for(int j = 1; j < d; j++)
            columns[i] += b[2*j][i] * b[2*j+1][i];
    }

    for(int i = 0; i < r_a; i++)
        for(int j = 0; j < c_b; j++) {
            c[i][j] = -rows[i] - columns[j];
            for(int k = 0; k < d; k++)
                c[i][j] += (a[i][2*k] + b[2*k+1][j]) * 
                            (a[i][2*k+1] + b[2*k][j]);
        }

    if(c_b%2)
        for(int i = 0; i < r_a; i++)
            for(int j = 0; j < c_b; j++)
                c[i][j] += a[i][c_a-1] * b[c_a-1][j];

    free(rows);
    free(columns);
}

int** allocateM(size_t r, size_t c)
{
    int** matrix = (int**)malloc(sizeof(int*) * r);
    if(matrix)
    //    for(int i = 0; i < r; i++)
    //        matrix[i] = (int*)malloc(sizeof(int) * c);
        matrix[0] = (int*)malloc(sizeof(int) * c * r);
    else
        return NULL;

    for(int i = 0; i < r; i++)
        matrix[i] = (*matrix + c * i);
    
    for(int i = 0; i < r; i++)
        for(int j = 0; j < c; j++)
            matrix[i][j] = 0;

    return matrix;
}

int randM(int** a, size_t r, size_t c)
{
    for(int i = 0; i < r; i++)
        for(int j = 0; j < c; j++)
            a[i][j] = rand() % 50;
}

void printM( int** a, size_t len1, size_t len2 )
{
    for(int i = 0; i < len1; i++) {
        printf("\n");
        for(int j = 0; j < len2; j++)
            printf("%d ", a[i][j]);
    }
}

void freeM(int** a, size_t r)
{
    free(a[0]);
    //for(int i = 0; i < r; i++)
    //    free(a[i]);
    free(a);
}

File: lab_02/winograd.py
import random as r
import os

def main():
    m = genRandMatrix(4, 5)
    n = genRandMatrix(5, 4)
    
    c = winograd(m, n)

    os.system("clear")
    printMatrix(m)
    printMatrix(n)
    printMatrix(c)
    
    return 1


def winograd(m, n):
    rows = []
    
    a = len(m)
    b = len(m[0])

    c = len(n[0])
    
    d = b // 2
    rows = [0 for i in range(a)]
    for i in range(a):
        rows[i] += m[i][0] * m[i][1]
        for j in range(1, d):
            rows[i] += m[i][2*j] * m[i][2*j+1]

    columns = [0 for i in range(c)]
    for i in range(c):
        columns[i] += n[0][i]*n[1][i]
        for j in range(1, d):
            columns[i] += n[2*j][i] * n[2*j+1][i]

    r = [[0 for i in range(c)] for j in range(a)]
    for i in range(a):
        for j in range(c):
            r[i][j] = - rows[i] - columns[j]
            for k in range(d):
                r[i][j] += (m[i][2*k] + n[2*k+1][j]) * (m[i][2*k+1] + n[2*k][j])

    if b%2:
        for i in range(a):
            for j in range(c):
                r[i][j] += m[i][b-1] * n[b-1][j]

    return r


def genRandMatrix(a, b):
    return [[r.randint(0, 20) for i in range(b)] for i in range(a)]


def printMatrix(a):
    if not a:
        return
    print("\n")
    for i in range(len(a)):
        for j in range(len(a[0])):
            print(" ", a[i][j], end="")
        print("\n")


if __name__ == "__main__":
    main()
